compiler/src/main/scala/org/scalajs/nscplugin/PrepJSExports.scala

```diff
@@ -548,30 +548,65 @@ trait PrepJSExports[G <: Global with Singleton] { this: PrepJSInterop[G] =>
 
       clsSym.info.decls.enter(expGetter)
 
-      genProxyDefDef(clsSym, trgGetter, expGetter, pos)
+      genProxyDefDef(trgGetter, expGetter, pos)
 
     } else EmptyTree
   }
 
   /** generate a DefDef tree (from [[proxySym]]) that calls [[trgSym]] */
-  private def genProxyDefDef(clsSym: Symbol, trgSym: Symbol,
-      proxySym: Symbol, pos: Position) = atPos(pos) {
-
-    // Helper to ascribe repeated argument lists when calling
-    def spliceParam(sym: Symbol) = {
-      if (isRepeated(sym))
-        Typed(Ident(sym), Ident(tpnme.WILDCARD_STAR))
-      else
-        Ident(sym)
+  private def genProxyDefDef(trgSym: Symbol, proxySym: Symbol, pos: Position) = atPos(pos) {
+    val fun = gen.mkAttributedRef(trgSym)
+
+    val nonPolyFun: Tree = fun.tpe match {
+      case PolyType(tparams, restpe) =>
+        val targs = proxySym.typeParams
+        val newRestpe =
+          restpe.instantiateTypeParams(tparams, targs.map(_.typeConstructor))
+        TypeApply(fun, targs.map(gen.mkAttributedIdent(_))).setType(newRestpe)
+
+      case _ =>
+        assert(proxySym.typeParams.isEmpty, s"got unexpected type params when proxying $trgSym")
+        fun
     }
 
-    // Construct proxied function call
-    val sel = Select(This(clsSym), trgSym)
-    val rhs = proxySym.paramss.foldLeft[Tree](sel) {
-      (fun,params) => Apply(fun, params map spliceParam)
+    val rhs = proxySym.paramss.foldLeft(nonPolyFun) { (fun, params) =>
+      fun.tpe match {
+        case MethodType(_, restpe) =>
+          val args = params.map { param =>
+            val ident = gen.mkAttributedIdent(param)
+
+            if (isRepeated(param)) Typed(ident, Ident(tpnme.WILDCARD_STAR))
+            else ident
+          }
+
+          Apply(fun, args).setType(restpe)
```

Comment:
Directly using `restpe` won't work here for dependent methods, i.e., methods where the result type, or param types in secondary param lists, use path-dependent types of the previous parameters.

---

In general, I find this new strategy dangerous. It doesn't correspond to how scalac usually works. All scalac phases create untyped trees that it passes to the `typer` to set the correct types. We can do better than what we did before by at least using the overloads of `gen.mkMethodCall`. But I wouldn't go as far as calling `setType` ourselves.

Reply:
> In general, I find this new strategy dangerous.

I agree. It essentially duplicates typer logic. Which is bad.

> It doesn't correspond to how scalac usually works. All scalac phases create untyped trees that it passes to the typer to set the correct types.

I was unaware of that. In this case, we definitely shouldn't do this. But then how do these phases "protect" themselves from crashing when calling typer?

Reply:
> But then how do these phases "protect" themselves from crashing when calling typer?

Mostly, they assume that because they received well-typed trees, the trees they produce will be well-typed as well.

This is probably the one thing that I find immensely better in Scala 3. There, it is the norm to create typed trees without going through untyped trees, and there are good internal APIs to do so.